forte/data/data_store.py

```diff
@@ -225,6 +229,223 @@ def __init__(
         """
         self.__deletion_count: dict = {}
 
+    def __getstate__(self):
+        r"""
+        In serialization,
+            1) will serialize the annotation sorted list as a normal list;
+            2) will remove `_onto_file_path`, `_dynamically_add_type`,
+                `tid_ref_dict`, `tid_idx_dict` and `deletion_count` to save space.
+        """
+        state = super().__getstate__()
+        for k in state["_DataStore__elements"]:
+            # build the full `_type_attributes`
+            self._get_type_info(k)
+            state["_DataStore__elements"][k] = list(
+                state["_DataStore__elements"][k]
+            )
+        state.pop("_onto_file_path")
+        state.pop("_dynamically_add_type")
+        state.pop("_DataStore__tid_ref_dict")
+        state.pop("_DataStore__tid_idx_dict")
+        state.pop("_DataStore__deletion_count")
+        state["entries"] = state.pop("_DataStore__elements")
+        state["fields"] = self._type_attributes
+        for _, v in state["fields"].items():
+            v.pop("parent_entry")
+        return state
+
+    def __setstate__(self, state):
+        r"""
+        In deserialization, we
+            1) transform the annotation list back to a sorted list;
+            2) recreate the `tid_ref_dict` and `tid_idx_dict` from `__elements`;
+            3) reset the `deletion_count`.
+        """
+        self.__dict__.update(state)
+
+        self.__elements = self.__dict__.pop("entries")
+        self._type_attributes = self.__dict__.pop("fields", {})
+        self._DataStore__tid_ref_dict = {}
+        self._DataStore__tid_idx_dict = {}
+        self._DataStore__deletion_count = {}
+
+        reset_index = {}
+        for k in self.__elements:
+            if self._is_annotation(k):
+                self.__elements[k] = SortedList(self.__elements[k])
+            for e in self.__elements[k]:
+                if e is not None:
+                    if self._is_annotation(k):
+                        # annotation-like
+                        self._DataStore__tid_ref_dict[
+                            e[constants.TID_INDEX]
+                        ] = e
+                    else:
+                        # non-annotation-like
+                        # use `reset_index` to recalculate indices
+                        type_name = e[constants.ENTRY_TYPE_INDEX]
+                        if type_name in reset_index:
+                            reset_index[type_name] += 1
+                        else:
+                            reset_index[type_name] = 1
```

Comment:
1. We need better documentation in this whole code block to explain what this is doing, otherwise it is very hard to maintain in the future
2. I still don't understand what `reset_index` is doing, it looks like a counter of how many times a particular type appears, why can this be used as the index for a particular entry instance?
3. [This](https://github.com/asyml/forte/blob/2af0617f5b9280e3caf2aa0f757b259a676697ef/forte/data/data_store.py#L208) is the definition of the `tid_idx_dict`, why are you using `type_name` as the key? The key should be `tid`, no?